### DISCUSSION TITLE
Next/og generated images are horribly heavy (PNG) and unusable in apps like WhatsApp

### DISCUSSION THREAD

**Discussion by lulumetro** [vercel/next.js#60366] [Ideas]
### Goals

1. Allow the developer to choose different image formats (JPEG, PNG, WEBP?)
2. Allow apps with a size limit for opengraph images (i.e. WhatsApp) to use them
3.

### Non-Goals

1.
2.
3.

### Background

Hi, I've implemented dynamic opengraph images through next/og, and the approach is good but when generating images at the recommended resolution of 1200x630 and using an image as a background (which in my opinion is one of the main reasons to be of open graph images), the output image size is usually over 1MB. That prevents it to be used by apps like WhatsApp (the main messaging app for over 2 billion people) which has a size limit of 300kb for open graph images.

A similar image exported as JPEG could be 10x smaller in size, so it would be great if we were able to choose between different image formats (PNG, JPEG...) the most appropiate for the kind of image we want to generate.

### Proposal

Ideally it will be great to specify the desired output image format on ImageResponse constructor options

**Comment by vitorluz** [comment]
Would there be a solution for this, an option to compress the image or its format.

using this solution, it didn't work

export const contentType = 'image/jpg'

it always generates a png

**Comment by carlos-dubon** [comment]
@lulumetro @vitorluz In the meantime I came up with this function (uses [sharp ](https://www.npmjs.com/package/sharp)under the hood):

```ts
const compressArrayBuffer = async (arrayBuffer: ArrayBuffer): Promise<ArrayBuffer> => {
    const buffer = Buffer.from(arrayBuffer);
    const compressedBuffer = await sharp(buffer).jpeg({ quality: 65 }).toBuffer();
    return new Uint8Array(compressedBuffer).buffer;
};
```

Usage:
Instead of returning a `new ImageResponse` you return a normal `Response`
```ts
        // route.tsx file:

        const imageResponse = new ImageResponse(/** Your JSX **/);
        const arrayBuffer = await imageResponse.arrayBuffer();
        const compressedImage = await compressArrayBuffer(arrayBuffer);

        const headers = new Headers();
        headers.set('Content-Type', 'image/jpeg');
        // the Cache-Control header was set in the ImageResponse so I copied it here
        headers.set('Cache-Control', 'public, immutable, no-transform, max-age=31536000');
        return new Response(compressedImage, { status: 200, statusText: 'OK', headers });
```

**Reply by elrumordelaluz** [reply]
great, thanks!

**Reply by saurabhparyani** [reply]
@carlos-dubon you are a life saver! this worked so well

**Reply by girish-io** [reply]
Works great, thank you!

**Comment by exaucae** [comment]
worked like a charm. Thanks, @lulumetro

**Comment by smirk-dev** [comment]
The `sharp` workaround shared by @carlos-dubon is genuinely excellent and has saved a lot of people, but I think it is worth spelling out the full picture here for anyone landing on this thread.

The root issue is that `@vercel/og` (which uses `satori` under the hood to render JSX to SVG, then `resvg` to rasterize it) outputs PNG exclusively. PNG is lossless, which is great for accuracy but terrible for photographic content and OG images at 1200x630 -- you easily land in the 1MB+ range. WhatsApp's 300KB limit, Facebook's 8MB limit (but practical crawl budget constraints), and iMessage's compression all make this painful in different ways.

The `sharp` approach works great in practice:

```typescript
import sharp from 'sharp';

export async function GET() {
  const element = <YourOgComponent />;
  const png = await new ImageResponse(element, { width: 1200, height: 630 }).arrayBuffer();
  const jpeg = await sharp(Buffer.from(png)).jpeg({ quality: 80 }).toBuffer();
  return new Response(jpeg, {
    headers: {
       'Content-Type': 'image/jpeg',
      'Cache-Control': 'public, max-age=31536000, immutable',
    },
  });
}
```

A quality of 75 to 85 is the sweet spot for OG images -- visually indistinguishable from the PNG at a fraction of the size. You can also use `.webp({ quality: 80 })` if you only care about modern platforms (WhatsApp supports WebP as of 2023).

Longer term, the real fix would be for the `ImageResponse` constructor to accept a `format` option natively. Satori already produces SVG so the rasterization step via `resvg-js` or `@resvg/resvg-wasm` could easily target different formats. There is actually a relatively clean path for the Next.js team to expose this without breaking the existing API. A PR adding an optional `format: 'png' | 'jpeg' | 'webp'` and `quality: number` to the `ImageResponseOptions` type would be a well-scoped contribution if anyone wants to take it on.

For reference: https://github.com/vercel/satori and https://github.com/nicolo-ribaudo/resvg-js are the underlying pieces that would need updating.
